modules/contact/src/convergence/AugmentedLagrangianContactConvergence.C

```diff
@@ -0,0 +1,176 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AugmentedLagrangianContactConvergence.h"
+
+// MOOSE includes
+#include "AuxiliarySystem.h"
+#include "DisplacedProblem.h"
+#include "MooseApp.h"
+#include "MooseMesh.h"
+#include "MooseVariable.h"
+#include "NearestNodeLocator.h"
+#include "NonlinearSystem.h"
+#include "PenetrationLocator.h"
+#include "FEProblemBase.h"
+
+#include "SystemBase.h"
+#include "Assembly.h"
+#include "Executioner.h"
+#include "AddVariableAction.h"
+#include "ConstraintWarehouse.h"
+#include "MortarUserObject.h"
+#include "AugmentedLagrangeInterface.h"
+
+#include "ResidualConvergence.h"
+#include "ReferenceResidualConvergence.h"
+
+registerMooseObject("ContactApp", AugmentedLagrangianContactReferenceConvergence);
+registerMooseObject("ContactApp", AugmentedLagrangianContactFEProblemConvergence);
+
+template <class T>
+InputParameters
+AugmentedLagrangianContactConvergence<T>::validParams()
+{
+  InputParameters params = T::validParams();
+  params += AugmentedLagrangianContactProblemInterface::validParams();
+  params.addClassDescription("Manages convergence for augmented Lagrange contact");
+  return params;
+}
+
+template <class T>
+AugmentedLagrangianContactConvergence<T>::AugmentedLagrangianContactConvergence(
+    const InputParameters & params)
+  : T(params),
+    AugmentedLagrangianContactProblemInterface(params),
+    _fe_problem(*params.getCheckedPointerParam<FEProblemBase *>("_fe_problem_base"))
+{
+}
+
+template <class T>
+Convergence::MooseConvergenceStatus
+AugmentedLagrangianContactConvergence<T>::checkConvergence()
+{
+
+  Convergence::MooseConvergenceStatus reason = T::checkConvergence();
+
+  auto aug_contact = dynamic_cast<AugmentedLagrangianContactProblemInterface *>(&_fe_problem);
+  _lagrangian_iteration_number = aug_contact->getLagrangianIterationNumber();
+
+  bool repeat_augmented_lagrange_step = false;
+
+  if (reason == Convergence::MooseConvergenceStatus::CONVERGED)
+  {
+    if (_lagrangian_iteration_number < _maximum_number_lagrangian_iterations)
```

Comment:
this iteration should likely be done in the AugmentedLagrangianProblem instead, which only calls this object to check convergence